### ISSUE TITLE
Add more highlights in docs and examples that coroutines definitions are always depend on loop context

### ISSUE THREAD

**Issue by rudyryk** [python/asyncio#332]
Hello everyone! This is more discussion issue than bug report or kind of.

For my opinion code examples in official asyncio docs are broken. Well, not totally broken of course, but they **give a wrong idea that coroutines definition are independent from loop context!**

I'd streess the word **definition**. Because I think it's clear, that coroutines **execution** depends on loop. But it's totally unclear, that **definition** also depends on it.

Here's an example from docs:  
https://docs.python.org/3/library/asyncio-task.html#example-coroutine-displaying-the-current-date

```python
import asyncio
import datetime

async def display_date(loop):
    end_time = loop.time() + 5.0
    while True:
        print(datetime.datetime.now())
        if (loop.time() + 1.0) >= end_time:
            break
        await asyncio.sleep(1)

loop = asyncio.get_event_loop()
# Blocking call which returns when the display_date() coroutine is done
loop.run_until_complete(display_date(loop))
loop.close()
```

It works.

Now let's change a single line:

```python
loop = asyncio.new_event_loop()
```

It doesn't.

But the fix isn't really hard:

```python
await asyncio.sleep(1, loop=loop)
```

Well, the next example is even worse:

```
import asyncio

async def compute(x, y):
    print("Compute %s + %s ..." % (x, y))
    await asyncio.sleep(1.0)
    return x + y

async def print_sum(x, y):
    result = await compute(x, y)
    print("%s + %s = %s" % (x, y, result))

loop = asyncio.get_event_loop()
loop.run_until_complete(print_sum(1, 2))
loop.close()
```

It can't run on custom loop. Because `compute()` and `print_sum()` coroutines have **no idea of event loop** and thus **can't be fixed without changing interface**.

Well, I think we should handle that really carefuly :)

Either rewrite examples in docs to provide correct patterns or handle custom loops automatically. Abound second, don't know if that's possible or really good idea :)

**Comment by Martiusweb** [comment]
Hi,

On the line :

        await asyncio.sleep(1)

The loop is implicitely retrievied from asyncio.get_event_loop().

However, when calling asyncio.new_event_loop(), it doesn't set the default
loop, returned by get_event_loop(). Thus, asyncio.sleep() is bound to a
different loop than the one executing the coroutine and will never
terminate.

If you want to use several loops, one can be the default loop returned by
get_event_loop(), the other one must be passed as an argument.

Also, when one awaits on a future/task which is not bound to the loop
executing the coroutine, an exception is raised. We may want
asyncio.sleep() to also raise an exception in that case.

**Comment by rudyryk** [comment]
Hi, @Martiusweb! Yes, I totally understand now, what happens and usually write my code against explicitly specified event loop.

But I came to that after **one year of development** with asyncio :) And now I can say, it's one of the most unclear and tricky parts. And so I find introductory docs confusing about that.

**Comment by rudyryk** [comment]
Well, while writing coroutine we should pick one of two options:

1. Enable support of arbitrary (not set as current) event loop
2. Always run on current event loop

The first option is more verbose and more explicit. The second one is compact and implicit. And that's not really simple question, which option if preferable in a particular case.

**Comment by Martiusweb** [comment]
`asyncio.sleep()` accepts an optional explicit loop argument, as any function or object constructor in `asyncio`. When the optional `loop` argument is `None`, the loop is retrieved from `asyncio.get_event_loop()`.

Indeed, reusable libraries should follow the convention of an optional loop argument and defaulting to `asyncio.get_event_loop()`.

In the application code, I can't see cases where you want to use several loops in one thread. I chose explicit loop passing most of the time, but one could use a custom `EventLoopPolicy` which sets a loop per thread, stored as a thread-local object.

**Comment by rudyryk** [comment]
I think it would be good to have such explanation as introduction to coroutines / tasks with code samples for both approaches :)

Currently, "Tasks and coroutines" documentation tells how to do things, but doesn't cover motivation and proper design practices. And I would insist that the [first example](https://docs.python.org/3/library/asyncio-task.html#example-coroutine-displaying-the-current-date) is broken :) If we pass loop as argument, but do not use it, it's probably a bug. Just imagine, that `display_date()` would be a 3rd party library.

**Comment by Martiusweb** [comment]
I agree: those examples are a bit misleading, and the documentation misses a general discussion about how and when asyncio should be used. I remember that the PEP (https://www.python.org/dev/peps/pep-3156/) helped me a bit.
